cmd/up.go

```diff
@@ -77,84 +80,114 @@ func Up() *cobra.Command {
 	}
 
 	namespace, client, restConfig, err := GetKubernetesClient(namespace)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	d, err := deployments.Get(namespace, dev.Swap.Deployment.Name, client)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
+
 	dev.Swap.Deployment.Container = deployments.GetDevContainerOrFirst(
 		dev.Swap.Deployment.Container,
 		d.Spec.Template.Spec.Containers,
 	)
+
 	devList, err := deployments.GetAndUpdateDevListFromAnnotation(d.GetObjectMeta(), dev)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	sy, err := syncthing.NewSyncthing(namespace, d.Name, devList)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	if err := deployments.DevModeOn(d, devList, client); err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	pod, err := deployments.GetCNDPod(ctx, d, client)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	go deployments.GetPodEvents(ctx, pod, client)
 
 	if err := deployments.InitVolumeWithTarball(ctx, client, restConfig, namespace, pod.Name, devList); err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	fullname := deployments.GetFullName(namespace, d.Name)
 
-	pf, err := forward.NewCNDPortForward(sy.RemoteAddress)
-	if err != nil {
-		return nil, err
-	}
-
 	if err := sy.Run(ctx, wg); err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	err = storage.Insert(ctx, wg, namespace, dev, sy.GUIAddress)
 	if err != nil {
 		if err == storage.ErrAlreadyRunning {
 			log.Infof("failed to insert new state value for %s", fullname)
-			return nil, fmt.Errorf("there is already an entry for %s. Are you running '%s up' somewhere else?", config.GetBinaryName(), fullname)
+			return nil, nil, fmt.Errorf("there is already an entry for %s. Are you running '%s up' somewhere else?", config.GetBinaryName(), fullname)
 		}
-		return nil, err
+		return nil, nil, err
+	}
+
+	pf, err := forward.NewCNDPortForward(sy.RemoteAddress)
+	if err != nil {
+		return nil, nil, err
 	}
 
 	if err := pf.Start(ctx, wg, client, restConfig, pod); err != nil {
-		return nil, fmt.Errorf("couldn't start the connection to your cluster: %s", err)
+		return nil, nil, fmt.Errorf("couldn't start the connection to your cluster: %s", err)
 	}
 
 	wg.Add(1)
 	go logs.StreamLogs(ctx, wg, d, dev.Swap.Deployment.Container, client)
 
 	go sy.Monitor(ctx, monitor)
-	return d, nil
+	return d, pf, nil
+}
+
+func reconnectPortForward(ctx context.Context, wg *sync.WaitGroup, d *appsv1.Deployment, pf *forward.CNDPortForward) error {
+
+	pf.Stop()
+
+	_, client, restConfig, err := GetKubernetesClient(d.Namespace)
+	if err != nil {
+		return err
+	}
+
+	pod, err := deployments.GetCNDPod(ctx, d, client)
+	if err != nil {
+		return err
+	}
+
+	if err := pf.Start(ctx, wg, client, restConfig, pod); err != nil {
+		return fmt.Errorf("couldn't start the connection to your cluster: %s", err)
```

Comment:
Since this is a reconnection `start` might be misleading. `Establish`? Is the user aware that it was trying to reconnect? What happens after this? Maybe the message should inform about the final result/impact rather of what happens to this particular piece.

Reply:
Yes, that's a good point. Let's do a separate PR review of those.